File: volume1/chapter7/ruby_sdk_example/ec2_example.rb
require 'rubygems'
require 'bundler/setup'
require 'aws-sdk-ec2'

ec2 = Aws::EC2::Client.new(region: 'us-east-1')

# Get image_id.
images_result = ec2.describe_images(
    {filters: [
        {name: 'description', values: ['Amazon Linux AMI 2018.03*']},
        {name: 'virtualization-type', values: ['hvm']}
    ]}
    )
image_id = images_result.images[0].image_id

# Get security_group_id.
sg_result = ec2.describe_security_groups(
    {filters: [
        {name: 'group-name', values: ['default']}
    ]}
    )
group_id = sg_result.security_groups[0].group_id

# Get vpc_id.
vpcs_result = ec2.describe_vpcs(
    {filters: [
        {name: 'isDefault', values: ['true']}
    ]}
    )
vpc_id = vpcs_result.vpcs[0].vpc_id

# Get subnet_id using previously computed vpc_id
subnets_result = ec2.describe_subnets(
    {filters: [
        {name: 'vpc-id', values: [vpc_id]}
    ]}
    )
subnet_id = subnets_result.subnets[0].subnet_id
puts "image_id=#{image_id} group_id=#{group_id} vpc_id=#{vpc_id} subnet_id=#{subnet_id}"

# Create instance
ec2_resource = Aws::EC2::Resource.new(region: 'us-east-1')
instance = ec2_resource.create_instances({
  image_id: image_id,
  subnet_id: subnet_id,
  security_group_ids: [group_id],
  instance_type: 't2.micro',  
  key_name: 'aws_intro_2',  
  min_count: 1,
  max_count: 1
})

instance_id = instance.first.id
puts "instance_id=#{instance_id}" 
instance.batch_create_tags({ tags: [{ key: 'Name', value: 'Created by ec2_example.rb' }]})
ec2_resource.client.wait_until(:instance_status_ok, {instance_ids: [instance_id]})

File: volume1/chapter7/go_sdk_example/ec2_example.go
package main

import (
    "fmt"

      "github.com/aws/aws-sdk-go/aws"
      "github.com/aws/aws-sdk-go/aws/session"
      "github.com/aws/aws-sdk-go/service/ec2"
)

func main() {
    // Load session from shared config   
    sess := session.Must(session.NewSessionWithOptions(session.Options{
        SharedConfigState: session.SharedConfigEnable,
    }))

    // Create new EC2 client
    ec2Svc := ec2.New(sess)

    // Get image_id
    image_id := ""
    images_input := &ec2.DescribeImagesInput{
		Filters: []*ec2.Filter{
			{
				Name:   aws.String("description"),
				Values: []*string{aws.String("Amazon Linux AMI 2018.03.0*")},
			},
			{
				Name:   aws.String("virtualization-type"),
				Values: []*string{aws.String("hvm")},
			},
		},
	}
    
    result, err := ec2Svc.DescribeImages(images_input)
    if err != nil {
        fmt.Println("Error", err)
    } else {
        image_id = *result.Images[0].ImageId
    } 

    // Get security_group_id
    group_id := ""
    group_input := &ec2.DescribeSecurityGroupsInput{
		Filters: []*ec2.Filter{
			{
				Name:   aws.String("group-name"),
				Values: []*string{aws.String("default")},
			},
		},
	}
    
    result_groups, err := ec2Svc.DescribeSecurityGroups(group_input)
    if err != nil {
        fmt.Println("Error", err)
    } else {
        group_id = *result_groups.SecurityGroups[0].GroupId
    }
 
    // Get vpc_id
    vpc_id := ""
    vpc_input := &ec2.DescribeVpcsInput{
		Filters: []*ec2.Filter{
			{
				Name:   aws.String("isDefault"),
				Values: []*string{aws.String("true")},
			},
		},
	}
    
    result_vpcs, err := ec2Svc.DescribeVpcs(vpc_input)
    if err != nil {
        fmt.Println("Error", err)
    } else {
        vpc_id = *result_vpcs.Vpcs[0].VpcId
    }
    
    // Get subnet_id from previously computed vpc_id
    subnet_id := ""
    subnet_input := &ec2.DescribeSubnetsInput{
		Filters: []*ec2.Filter{
			{
				Name:   aws.String("vpc-id"),
				Values: []*string{aws.String(vpc_id)},
			},
		},
	}
    
    result_subnets, err := ec2Svc.DescribeSubnets(subnet_input)
    if err != nil {
        fmt.Println("Error", err)
    } else {
        subnet_id = *result_subnets.Subnets[0].SubnetId
    }
    fmt.Println("image_id=", image_id, 
        "group_id=", group_id,
        "vpc_id=", vpc_id,
        "subnet_id=", subnet_id,
    )     
    
    instance_id := ""
    run_input := &ec2.RunInstancesInput{
        ImageId: aws.String(image_id),
        SubnetId: aws.String(subnet_id),
        SecurityGroupIds: []*string {aws.String(group_id)},
        InstanceType: aws.String("t2.micro"),
        KeyName: aws.String("aws_intro_2"),
        MinCount: aws.Int64(1),
        MaxCount: aws.Int64(1),
	}
    
    reservation, err := ec2Svc.RunInstances(run_input)
    if err != nil {
        fmt.Println("Error", err)
    } else {
        instance_id = *reservation.Instances[0].InstanceId
    }
    fmt.Println("instance_id", instance_id)  
    
    ec2Svc.WaitUntilInstanceRunning(&ec2.DescribeInstancesInput{
        InstanceIds: []*string{aws.String(instance_id)},
    })  
}
